src/Features/CSharp/Portable/ConvertIfToSwitch/CSharpConvertIfToSwitchCodeRefactoringProvider.cs

```diff
@@ -1,214 +1,101 @@
+                return ReturnStatement(
+                    SwitchExpression((ExpressionSyntax)target, SeparatedList(sections.Select(AsSwitchExpressionArmSyntax))));
             }
 
-            // We do not offer a fix if the if-statement contains a break-statement, e.g.
-            //
-            //      while (...)
-            //      {
-            //          if (...) {
-            //              break;
-            //          }
-            //      }
-            // 
-            // When the 'break' moves into the switch, it will have different flow control impact.
-            protected override bool CanConvertIfToSwitch(IfStatementSyntax ifStatement)
-                => !_semanticModel.AnalyzeControlFlow(ifStatement).ExitPoints.Any(n => n.IsKind(SyntaxKind.BreakStatement));
-
-            protected override IEnumerable<ExpressionSyntax> GetLogicalOrOperands(
-                ExpressionSyntax syntaxNode)
+            private static SwitchExpressionArmSyntax AsSwitchExpressionArmSyntax(SwitchSection section)
             {
-                syntaxNode = syntaxNode.WalkDownParentheses();
-                while (syntaxNode.IsKind(SyntaxKind.LogicalOrExpression))
-                {
-                    var binaryExpression = (BinaryExpressionSyntax)syntaxNode;
-                    yield return binaryExpression.Right.WalkDownParentheses();
-                    syntaxNode = binaryExpression.Left.WalkDownParentheses();
-                }
-
-                yield return syntaxNode;
+                return SwitchExpressionArm(
+                    pattern: section.Labels.IsDefault
+                        ? DiscardPattern()
+                        : AsPatternSyntax(section.Labels[0].Pattern),
+                    whenClause: section.Labels.IsDefault
+                        ? null 
+                        : AsWhenClause(section.Labels[0]),
+                    expression: AsExpressionSyntax(section.Body));
             }
 
-            protected override IEnumerable<(ExpressionSyntax, StatementSyntax)> GetIfElseStatementChain(IfStatementSyntax currentStatement)
+            private static ExpressionSyntax AsExpressionSyntax(IOperation operation)
             {
-                StatementSyntax elseClauseStatement;
-                do
-                {
-                    yield return (currentStatement.Condition, currentStatement.Statement);
-                    elseClauseStatement = currentStatement.Else?.Statement;
-                    currentStatement = elseClauseStatement as IfStatementSyntax;
-                }
-                while (currentStatement != null);
-
-                if (elseClauseStatement != null)
+                return operation switch
                 {
-                    yield return (null, elseClauseStatement);
-                }
+                    IReturnOperation op => (ExpressionSyntax)op.ReturnedValue.Syntax,
+                    IThrowOperation op => ThrowExpression((ExpressionSyntax)op.Exception.Syntax),
```

Comment:
fwiw i'm always slightly wary about trusting the nodes taht an operation points at...  but i'm not going to say tht you're disallowed from doing this (yet).

Reply:
Yeah I had that in mind when moving this to IOperation, although, operation nodes are pretty deterministic on what kind of syntax they point to. Maybe we can add a lang-specific validation layer to ensure they are of the expected type. but so far it worked just fine -- actually I didnt change any code while testing for VB.